alf/algorithms/misc_algorithm.py

```diff
@@ -0,0 +1,263 @@
+        self._empowerment = empowerment
+        self._env_name = env_name
+
+    def split_observation_tf(self, o):
+
+        dimo = o.get_shape().as_list()[-1]
+
+        if self._env_name in ['SocialBot-PlayGround-v0']:
+            if dimo == 21:
+                task_specific_ob, agent_pose, agent_vel, internal_states, action = tf.split(
+                    o, [3, 6, 6, 4, 2], axis=-1)
+
+                agent_pose_1, agent_pose_2 = tf.split(
+                    agent_pose, [3, 3], axis=-1)
+                joint_pose_1, joint_pose_2, joint_vel_1, joint_vel_2 = tf.split(
+                    internal_states, [1, 1, 1, 1], axis=-1)
+                joint_1 = tf.concat([joint_pose_1, joint_vel_1], axis=-1)
+                joint_2 = tf.concat([joint_pose_2, joint_vel_2], axis=-1)
+
+                if self._n_objects == 1:
+                    obs_achieved_goal = task_specific_ob
+                    obs_excludes_goal = agent_pose_1
+                    if self._empowerment:
+                        obs_excludes_goal = action
+                elif self._n_objects == 0:
+                    obs_achieved_goal = joint_1
+                    obs_excludes_goal = joint_2
+
+                return (obs_excludes_goal, obs_achieved_goal)
+
+            elif (dimo == 24) and (self._n_objects == 2):
+                task_specific_ob_1, task_specific_ob_2, agent_pose, agent_vel, internal_states, action = tf.split(
+                    o, [3, 3, 6, 6, 4, 2], axis=-1)
+
+                agent_pose_1, agent_pose_2 = tf.split(
+                    agent_pose, [3, 3], axis=-1)
+
+                obs_achieved_goal_1 = task_specific_ob_1
+                obs_achieved_goal_2 = task_specific_ob_2
+                obs_excludes_goal = agent_pose_1
+
+                return (obs_excludes_goal, obs_achieved_goal_1,
+                        obs_achieved_goal_2)
+            else:
+                print(
+                    'Please specify the state of interests and the context state in the misc_alorighm.py first.'
+                )
+                exit()
+        else:
+            print(
+                'Please specify the state of interests and the context state in the misc_alorighm.py first.'
+            )
+            exit()
+
+    def mine(self, x_in, y_in):
+        """Mutual Infomation Neural Estimator.
+
+        Implement mutual information neural estimator from
+        Belghazi et al "Mutual Information Neural Estimation"
+        http://proceedings.mlr.press/v80/belghazi18a/belghazi18a.pdf
+        'DV':  sup_T E_P(T) - log E_Q(exp(T))
+        where P is the joint distribution of X and Y, and Q is the product marginal distribution of P. DV is a lower bound for KLD(P||Q)=MI(X, Y).
+
+        """
+        y_in_tran = tf.transpose(y_in, perm=[1, 0, 2])
+        y_shuffle_tran = tf.gather(
+            y_in_tran, tf.random.shuffle(tf.range(
+                tf.shape(y_in_tran)
+                [0])))  # here tf.random.shuffle(y_in_tran) does not work
```

Comment:
Be more specific:  does not work because ...

Reply:
because it reports an error.

Reply:
use as required